keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,391 @@
+
+The idea of this KEP is to propose that namespaces are always ensured, by the API server, to have a *immutable label matching the metadata.name of the namespace*, which thus allows one to build label selectors to include/exclude namespaces by name, without having write access to a namespace (which is required to label it manually). 
+
+This was initially envisioned as a workaround to a limitation in the network policy API, but has emerged as a potential to other problems which may be solved by default namespace labels.
+
+## Motivation
+
+In https://github.com/kubernetes/enhancements/issues/2112 , we discussed several 
+ways to implement network policies that select namespaces by name, rather than by 
+relying on labels which may be unknown, or uneditable by people writing policies, 
+or untrustworthy.
+
+This was inspired by a broad community ask to target namespaces in this manner, 
+as evidenced by issues such as [kubernetes #88253](https://github.com/kubernetes/kubernetes/issues/88253), 
+and the larger set of user feedback obtained in  [user stories](https://github.com/jayunit100/network-policy-subproject/blob/master/p0_user_stories.md).
+
+It’s also worth noting that there are at least 2 other NamespaceSelector APIs in 
+k/k - Validating and Mutating webhooks.  They will inevitably need similar API 
+evolution, but we don't need to do that as part of this KEP.
+
+It turns out that alternate [solutions](#alternatives):
+Break the old API in fundamental ways that are questionably WRT cost/benefit OR
+add nesting complexity to the network policy API, wherein certain “subsets” of
+the API were not usable together.  This is not “bad”, but it is suboptimal and
+introduces redundant semantics. This is avoidable if we could assume, in all
+cases, that namespaces have a default label, thereby leveraging existing
+functionalities offered by label selectors.
+
+There may also be many other usecases where the security or permissions boundary
+might be conveniently defined in terms of a namepsace name, and that such a
+definition would be most easily referenced by a default label.
+
+### Goals
+
+- Add an immutable default label to ALL namespaces, as the namespace name so this 
+can be used as selector by arbitrary components.
+
+### Non-Goals
+
+* Publishing name labels on all resources
+
+Not all resource names satisfy the label requirements. Some of those resources require
+a name that can be used as a DNS subdomain name.
+* Exposing arbitrary fields as selectors
+
+## Proposal
+
+We propose to label all namespaces, by default, with a reserved label of
+“kubernetes.io/metadata.name” and set it to the namespace's name. This will allow
+any resource with a `namespaceSelector` field (such as NetworkPolicy, MutatingWebhookConfiguration
+etc) to select (or exclude with `notIn`) these namespaces by explicitly matching on
+the namespace's name with traditional label selector mechanisms.
```

Comment:
Are we confident that all the namespace label selectors we care about use `metav1.LabelSelector` rather than `map[string]string`? The problem with the latter is that under this proposal it would only be able to match a single namespace at a time.

Reply:
All three are `metav1.LabelSelector`